libraries/clchain/include/clchain/subchain.hpp

```diff
@@ -0,0 +1,201 @@
+#pragma once
+
+#include <clchain/graphql_connection.hpp>
+#include <eosio/bytes.hpp>
+#include <eosio/fixed_bytes.hpp>
+#include <eosio/name.hpp>
+#include <eosio/time.hpp>
+
+namespace subchain
+{
+   struct action
+   {
+      uint64_t seq = 0;
+      eosio::name firstReceiver;
+      eosio::name receiver;
+      eosio::name name;
+      eosio::bytes hexData;
+   };
+   EOSIO_REFLECT(action, seq, firstReceiver, receiver, name, hexData)
+
+   struct transaction
+   {
+      eosio::checksum256 id;
+      std::vector<action> actions;
+   };
+   EOSIO_REFLECT(transaction, id, actions)
+
+   struct eosio_block
+   {
+      uint32_t num = 0;
+      eosio::checksum256 id;
+      eosio::checksum256 previous;
+      eosio::time_point timestamp;
+      std::vector<transaction> transactions;
+   };
+   EOSIO_REFLECT(eosio_block, num, id, previous, timestamp, transactions)
+
+   struct block
+   {
+      uint32_t num = 0;
+      eosio::checksum256 previous;
+      eosio_block eosioBlock;
+   };
+   EOSIO_REFLECT(block, num, previous, eosioBlock)
+
+   struct block_with_id : block
+   {
+      eosio::checksum256 id;
+   };
+   EOSIO_REFLECT(block_with_id, id, base block)
+
+   inline const block& deref(const block& block) { return block; }
+   inline const block& deref(const std::unique_ptr<block>& block) { return *block; }
+   inline uint32_t get_block_num(uint32_t num) { return num; }
+   inline uint32_t get_block_num(const auto& block) { return deref(block).num; }
+   inline constexpr auto by_block_num = [](const auto& a, const auto& b) {
+      return get_block_num(a) < get_block_num(b);
+   };
+
+   inline const block_with_id& deref(const block_with_id& block) { return block; }
+   inline const block_with_id& deref(const std::unique_ptr<block_with_id>& block) { return *block; }
+   inline uint32_t get_eosio_num(uint32_t num) { return num; }
+   inline uint32_t get_eosio_num(const auto& block) { return deref(block).eosioBlock.num; }
+   inline constexpr auto by_eosio_num = [](const auto& a, const auto& b) {
+      return get_eosio_num(a) < get_eosio_num(b);
+   };
+
+   struct block_log
+   {
+      enum status
+      {
+         appended,
+         duplicate,
+         unlinkable,
+      };
+
+      static constexpr const char* status_str[] = {
+          "appended",
+          "duplicate",
+          "unlinkable",
+      };
+
+      std::vector<std::unique_ptr<block_with_id>> blocks;
+      uint32_t irreversible = 0;
+
+      const block_with_id* head() const
+      {
+         if (blocks.empty())
+            return nullptr;
+         return &*blocks.back();
+      }
+
+      const block_with_id* block_by_num(uint32_t num) const
+      {
+         auto it = std::lower_bound(blocks.begin(), blocks.end(), num, by_block_num);
```

Comment:
Since `add_block` requires the blocks to have sequential block numbers, a binary search is a waste of time.  You can just calculate the index.

Reply:
changed